sky/serve/serve_utils.py

```diff
@@ -891,14 +920,33 @@ def _get_replica_status() -> serve_state.ReplicaStatus:
 
     replica_provisioned = (
         lambda: _get_replica_status() != serve_state.ReplicaStatus.PROVISIONING)
-    with open(launch_log_file_name, 'r', newline='', encoding='utf-8') as f:
-        for line in _follow_logs_with_provision_expanding(
-                f,
-                replica_cluster_name,
+
+    # Handle launch logs based on number parameter
+    if tail is not None:
+        if tail - total_lines_printed <= 0:
+            return ''
+        static_lines = common_utils.read_last_n_lines(
+            launch_log_file_name, tail - total_lines_printed)
+        temp_file = io.StringIO(''.join(static_lines))
```

Comment:
This approach feels a bit indirect, as we get a list of lines from `read_last_n_lines` only to join it back into a stream to be parsed. What if we extracted the 'provision expanding' logic into its own helper function that could be applied directly to the list of lines?